File: archived/src/python/ch01/1.2/l2/karatsuba.py
def karatsuba(x: int, y: int) -> int:
    """Recursive Karatsuba multiplication."""
    if x < 10 or y < 10:
        return x * y

    n = max(x.bit_length(), y.bit_length())
    m = n // 2

    high1, low1 = divmod(x, 1 << m)
    high2, low2 = divmod(y, 1 << m)

    z0 = karatsuba(low1, low2)
    z2 = karatsuba(high1, high2)
    z1 = karatsuba(low1 + high1, low2 + high2) - z0 - z2

    return (z2 << (2 * m)) + (z1 << m) + z0

def main():
    a, b = 1234, 5678
    print("Karatsuba result:", karatsuba(a, b))
    print("Expected:", a * b)

if __name__ == "__main__":
    main()

File: archived/v1/src/c/ch01/1.2/l2/karatsuba.c
#include "karatsuba.h"

long long karatsuba(long long x, long long y) {
    if (x < 10 || y < 10) return x * y;

    int n = (x > y ? x : y);
    int bits = 0;
    while (n > 0) {
        bits++;
        n >>= 1;
    }
    int m = bits / 2;

    long long high1 = x >> m;
    long long low1  = x - (high1 << m);
    long long high2 = y >> m;
    long long low2  = y - (high2 << m);

    long long z0 = karatsuba(low1, low2);
    long long z2 = karatsuba(high1, high2);
    long long z1 = karatsuba(low1 + high1, low2 + high2) - z0 - z2;

    return (z2 << (2 * m)) + (z1 << m) + z0;
}
